syntax = "proto3";

message CreateUserRequest {

    string username = 1;
    string password = 2;
    string address = 3;

}

message CreateUserResponse {
    int32 response_code = 1;

    /*

    0 - Success
    1 - Username already exists
    2 - Invalid user parameters

    */
}

message AuthenticateUserRequest {
    string username = 1;
    string password = 2;
}

message AuthenticateUserResponse {

    int32 response_code = 1;
    optional string user_id = 2;

    /*

    0 - Success
    1 - User not found
    2 - Incorrect password

    */

}

message GetUserDetailsRequest {

    string user_id = 1;

}

message GetUserDetailsResponse {

    int32 response_code = 1;
    optional string username = 2;
    optional string address = 3;

    /*

    0 - Success
    1 - User not found

    */

}

message PayCartRequest {

    string user_id = 1;

}

message Invoice {

    int64 invoice_id = 1;
    double price = 2;
    int64 order_id = 3;
    int64 customer_id = 4;
    string fiscal_address = 5;
    optional string details = 6;


}

message PayCartResponse {

    int32 response_code = 1;
    optional Invoice invoice = 2;

    /*

    0 - Success
    1 - User not found
    2 - Cart is empty
    3 - Invalid Order

    */

}

message AddItemCartRequest {
    
    string user_id = 1;
    int64 batch_id = 2;
    double volume = 3;
}

message AddItemCartResponse {
    int32 response_code = 1;

    /*
    0 - Success
    1 - User was not found
    2 - Item is not valid
    3 - Item volume is not valid
    */

}

message DeleteItemCartRequest {
    string user_id = 1;
    optional int64 batch_id = 2;
}

message DeleteItemCartResponse {
    int32 response_code = 1;

    /*
    0 - Success
    1 - User was not found
    2 - Item is not in the cart
    */
}

message GetCartContentRequest {

    string user_id = 1;

}

message CartContent {
        int64 batch_id = 1;
        double volume = 2;
    }

message GetCartContentResponse {

    int32 response_code = 1;
    repeated CartContent content = 2;

    /*
    0 - Success
    1 - User was not found
    */

}

service User {
    rpc CreateUser (CreateUserRequest) returns (CreateUserResponse);
    rpc AuthenticateUser (AuthenticateUserRequest) returns (AuthenticateUserResponse);
    rpc GetUser (GetUserDetailsRequest) returns (GetUserDetailsResponse);
    rpc PayCart (PayCartRequest) returns (PayCartResponse);
    rpc AddItemCart (AddItemCartRequest) returns (AddItemCartResponse);
    rpc DeleteItemCart (DeleteItemCartRequest) returns (DeleteItemCartResponse);
    rpc GetCartContent (GetCartContentRequest) returns (GetCartContentResponse);
}